src/game_events/handlers.cpp

```diff
@@ -241,10 +241,29 @@ struct filter_attack : public event_filter {
 	{
 		const unit_map& units = resources::gameboard->units();
 		const auto& loc = first_ ? event_info.loc1 : event_info.loc2;
-		auto unit = units.find(loc);
-		if(unit != units.end() && loc.matches_unit(unit)) {
-			const config& attack = event_info.data.child(first_ ? "first" : "second");
-			return swf_.empty() || matches_special_filter(attack, swf_);
+		const auto& loc_d = first_ ? event_info.loc2 : event_info.loc1;
+		auto unit_a = units.find(loc);
+		auto unit_d = units.find(loc_d);
+		if(unit_a != units.end() && loc.matches_unit(unit_a)) {
+			const unit& u = *unit_a;
+			const config& temp_weapon = event_info.data.child(first_ ? "first" : "second");
+			auto weapon = std::make_shared<const attack_type>(temp_weapon);
+			const_attack_ptr attack = weapon->shared_from_this();
+			if(unit_d != units.end() && loc_d.matches_unit(unit_d)) {
+				const unit& opp = *unit_d;
+				const config& temp_other_weapon = event_info.data.child(!first_ ? "first" : "second");
+				auto opp_weapon = std::make_shared<const attack_type>(temp_other_weapon);
+				const_attack_ptr second_attack = opp_weapon->shared_from_this();
+				auto ctx = attack->specials_context(u.shared_from_this(), opp.shared_from_this(), loc, loc_d, first_, second_attack);
+				std::optional<decltype(ctx)> opp_ctx;
+				if(second_attack) {
+					opp_ctx.emplace(second_attack->specials_context(opp.shared_from_this(), u.shared_from_this(), loc_d, loc, !first_, attack));
+				}
+				return swf_.empty() || matches_special_filter(attack, swf_ , loc);
+			} else {
+			auto ctx = attack->specials_context(u.shared_from_this(), loc, first_);
```

Comment:
What's this for? It seems to go out of scope before being used.

Also, the indentation on this line is wrong.

Reply:
i fix that(i think)

Reply:
Ok, indentation is fixed and the "goes out of scope" was my misunderstanding (the context is accessible via the weapon, and removes itself when it goes out of scope).